### PR TITLE
Plant support for hard forks

### PR DESCRIPTION
**Pull request by lucafavatella**
This PR aims at finishing https://www.pivotaltracker.com/story/show/153778514

* Block validation is system tested with chain with old protocol - not new
* Generation of block candidate and mining are not tested
  * I plan to integrate system test for mining coinbase
  * I do not plan to integrate with txs different from coinbase (I may smoke test with spend tx)
    * While reviewing you are welcome to reach protocol switch height and test sending txs - it shall work

TODO:
* [ ] Test new node mines with coinbase only, and subsequently syncs with other new node (hence passing validation with both old and new protocol)
* [ ] Test against latest UAT/testnet DB snapshot
* [ ] Mention in release notes process in wiki that before tagging height of new protocol needs to be changed
* [ ] Rebase on master

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -152,7 +152,18 @@ eunit:
...
+system-test-deps: | system_test/aest_hard_fork_SUITE_data/db/mnesia_ae-uat-epoch_backup.tar ;
+
+system_test/aest_hard_fork_SUITE_data/db/mnesia_ae-uat-epoch_backup.tar: system_test/aest_hard_fork_SUITE_data/db/mnesia_ae-uat-epoch_backup
+	tar -c -C $(<D) -f $@ $(<F)
+
+system_test/aest_hard_fork_SUITE_data/db/mnesia_ae-uat-epoch_backup: system_test/aest_hard_fork_SUITE_data/db/mnesia_ae-uat-epoch_backup.gz
+	zcat <$< >$@
+
+system_test/aest_hard_fork_SUITE_data/db/mnesia_ae-uat-epoch_backup.gz:
+	curl -fsS --create-dirs -o $@ https://6798-99802036-gh.circle-artifacts.com/0/tmp/chain_snapshots/54.202.11.197/tmp/mnesia_ip-172-31-36-149_db_backup_1523145687.gz
```
Comment (lucafavatella) [inline Makefile]: @acidsennui Here I retrieve the DB snapshot with the chain. It is hardcoded link, I think it is fine for the moment (though I plan to try the newer snapshot on CI in scope of this PR).

File: apps/aecore/src/aec_governance.erl [hunk 2]
```diff
@@ -13,11 +15,40 @@
...
          name_registrars/0]).
 
+-export_type([protocols/0]).
+
+-include("common.hrl").
+-include("blocks.hrl").
+
+-define(SORTED_VERSIONS, lists:sort(maps:keys(?PROTOCOLS))).
+-define(PROTOCOLS, #{?GENESIS_VERSION => ?GENESIS_HEIGHT,
+                     ?PROTOCOL_VERSION => 5000}). %% TODO Adapt before release.
```
Comment (lucafavatella) [inline apps/aecore/src/aec_governance.erl]: This is the shipped height. (It is configurable for testing.)

File: system_test/aest_hard_fork_SUITE.erl [hunk 3]
```diff
@@ -0,0 +1,334 @@
...
+-export([init_per_suite/1]).
+-export([end_per_suite/1]).
+-export([init_per_group/2]).
+-export([end_per_group/2]).
+-export([init_per_testcase/2]).
+-export([end_per_testcase/2]).
+
+% Test cases
+-export([
+         node_persisting_chain_and_not_mining_has_genesis_as_top/1,
```
Comment (jur0) [inline system_test/aest_hard_fork_SUITE.erl]: I'd prefer the test case name to explicitly show which node is concerned in the test. In this case it's `old_node_persisting_chain_and_not_mining_has_genesis_as_top`.

**Comment by jur0** [comment]
The following test case, that checks if the new node can sync from the old node (not just from db backup), should be included:

```erlang
new_node_can_receive_short_old_chain_from_old_node(Cfg) ->
    {_, TopHeight} = proplists:lookup(db_backup_top_height, Cfg),
    {_, TopHash} = proplists:lookup(db_backup_top_hash, Cfg),
    aest_nodes:setup_nodes([?OLD_NODE1, ?OLD_NODE2, ?NEW_NODE1], Cfg),
    start_node(old_node1, Cfg),
    {ok, {TopHash, TopHeight}} = restore_db_backup_on_node(old_node1, Cfg),
    start_node(new_node1, Cfg),
    aest_nodes:wait_for_height(TopHeight, [old_node1], 5000, Cfg),
    wait_for_height_syncing(TopHeight, [new_node1], {{45000, ms}, {200, blocks}}, Cfg),
    B = get_block_by_height(new_node1, TopHeight, Cfg),
    ?assertEqual(TopHash, maps:get(hash, B)),
    ok.
```

**Comment by lucafavatella** [comment]
> The following test case, that checks if the new node can sync from the old node (not just from db backup), should be included:
>
> new_node_can_receive_short_old_chain_from_old_node(Cfg) ->

@jur0 Peer-to-peer networking in (to-be-released) v0.11.0 is meant not to be backwards compatible with v0.10.1, hence this test shall not be added in this release cycle. BTW this is why I test that new node can start from *short* old backup - so to have a way of seeding old chain in new peer-to-peer-network (of 0.11.0 nodes).

File: system_test/aest_hard_fork_SUITE.erl (hunk 3 above)
Comment (lucafavatella) [reply system_test/aest_hard_fork_SUITE.erl]: Addressed in d641ba52

File: system_test/aest_hard_fork_SUITE.erl [hunk 4]
```diff
@@ -0,0 +1,334 @@
...
+         restore_db_backup_on_same_release/1,
```
[the remark this answers is not in the document]
Comment (lucafavatella) [reply system_test/aest_hard_fork_SUITE.erl]: Addressed in d641ba52

File: system_test/aest_hard_fork_SUITE.erl [hunk 5]
```diff
@@ -0,0 +1,334 @@
...
+         old_node_can_receive_chain_from_other_old_node/1,
+         restore_db_backup_with_short_chain_on_new_release/1,
```
[the remark this answers is not in the document]
Comment (lucafavatella) [reply system_test/aest_hard_fork_SUITE.erl]: Addressed in d641ba52

**Comment by lucafavatella** [comment]
See #967 - with mining tests and rebased on master.

**Pull request closed without merging**